### PR TITLE
Grouped query attention

### PR DESCRIPTION
**Pull request by drisspg**
# Summary
Closed: #106730 in favor of this PR since the rebase would have been more difficult then just recreating.

There is still some edge cases I need to appropriately handle but this is how it could look.#109887 

## Design choices
This is currently implemented by looking at the "num heads" dimension, size(1) and determining if it can be repeat_interleaved to fill the query num_heads. 

Another option would be to pass in an explicit `groups` parameter to the top level SDPA and passing down.

### CODE REVIEW

File: aten/src/ATen/native/transformers/attention.cpp [hunk 1]
```diff
@@ -12,6 +12,7 @@
 #include <ATen/native/transformers/attention.h>
 #include <ATen/native/transformers/sdp_utils_cpp.h>
 #include <utility>
+#include "ATen/ops/repeat_interleave.h"
```
Comment (drisspg) [inline aten/src/ATen/native/transformers/attention.cpp:15]: remove

**Review by cpuhrsch — CHANGES_REQUESTED** [review]
Please use an explicit keyword argument to enable this behavior.

We can always switch that to `True` by default, but right now your operator violates PyTorch broadcasting principles.

**Comment by Muennighoff** [comment]
@drisspg Does the new PyTorch 2.2 support GQA / MQA? As this PR was closed, are there plans for adding it in the future if not?

**Pull request closed without merging**